cylc/flow/config.py

```diff
@@ -1008,6 +1017,220 @@ def _check_sequence_bounds(self):
+        try:
+            outputs = self.taskdefs[task_name].outputs
+        except KeyError:
+            # this is a family -> we'll check integrity for each task that
+            # inherits from it
+            return
+
+        (
+            _trigger_to_completion_variable,
+            _completion_variable_to_trigger,
+        ) = get_trigger_completion_variable_maps(outputs.keys())
+
+        # get the optional/required outputs defined in the graph
+        graph_optionals = {
+            # completion_variable: is_optional
+            _trigger_to_completion_variable[trigger]: (
+                None if is_required is None else not is_required
+            )
+            for trigger, (_, is_required)
+            in outputs.items()
+        }
+        if (
+            graph_optionals[TASK_OUTPUT_SUCCEEDED] is True
+            and graph_optionals[TASK_OUTPUT_FAILED] is None
+        ):
+            # failed is implicitly optional if succeeded is optional
+            # https://github.com/cylc/cylc-flow/pull/6046#issuecomment-2059266086
+            graph_optionals[TASK_OUTPUT_FAILED] = True
+
+        # get the optional/required outputs defined in the expression
+        try:
+            # this involves running the expression which also validates it
+            expression_optionals = get_optional_outputs(expr, outputs)
+        except NameError as exc:
+            # expression references an output which has not been registered
+            error = exc.args[0][5:]
+
+            if f"'{TASK_OUTPUT_FINISHED}'" in error:
+                # the finished output cannot be used in completion expressions
+                # see proposal point 5::
+                # https://cylc.github.io/cylc-admin/proposal-optional-output-extension.html#proposal
+                raise WorkflowConfigError(
+                    f'Error in [runtime][{task_name}]completion:'
+                    f'\n  {expr}'
+                    '\nThe "finished" output cannot be used in completion'
+                    ' expressions, use "succeeded or failed".'
+                )
+
+            for alt_qualifier, qualifier in ALT_QUALIFIERS.items():
+                _alt_compvar = trigger_to_completion_variable(alt_qualifier)
+                _compvar = trigger_to_completion_variable(qualifier)
+                if re.search(rf'\b{_alt_compvar}\b', error):
+                    raise WorkflowConfigError(
+                        f'Error in [runtime][{task_name}]completion:'
+                        f'\n  {expr}'
+                        f'\nUse "{_compvar}" not "{_alt_compvar}" '
+                        'in completion expressions.'
+                    )
+
+            raise WorkflowConfigError(
+                # NOTE: str(exc) == "name 'x' is not defined" tested in
+                # tests/integration/test_optional_outputs.py
+                f'Error in [runtime][{task_name}]completion:'
+                f'\nInput {error}'
+            )
```

Comment:
Hmmm this results in declared but undefined outputs being flagged as Inputs, which is likely to confuse users:
```ini
[scheduling]
    [[graph]]
        R1 = "foo:x"
[runtime]
    [[foo]]
        completion = succeeded or x
        #[[[outputs]]]
        #   x = x
```
```
 $ cylc val .
WorkflowConfigError: Error in [runtime][foo]completion:
Input 'x' is not defined
```

Reply:
(I'll just delete "Input" there for now).